### PR TITLE
Review of fixed source bank model developments

### PR DESCRIPTION
**Pull request by davidjohnlong**

### CODE REVIEW

File: include/openmc/particle.h [hunk 1]
```diff
@@ -181,6 +181,15 @@ class Particle {
     int delayed_group; //!< particle delayed group
   };
 
+  //! Saved ("banked") state of primary source particle, for particle restart
+  struct RestartBank {
+	    Position r;
+	    Direction u;
+	    double E;
+	    double wgt;
```
Comment (simondrichards) [inline include/openmc/particle.h]: The OpenMC Style Guide for C++ says to use two spaces per indentation level. The other struct definitions in this file have conformed to this guide, so it's probably best if this one does too.

File: src/simulation.cpp [hunk 2]
```diff
@@ -105,7 +106,16 @@ int openmc_simulation_init()
     load_state_point();
     write_message("Resuming simulation...", 6);
   } else {
-    initialize_source();
+	//Only initialize primary source bank for eigenvalue simulations
+	if (settings::run_mode == RunMode::EIGENVALUE) {
+	  initialize_source();
+	}
```
Comment (simondrichards) [inline src/simulation.cpp]: Use two spaces per indentation level.

File: src/simulation.cpp [hunk 3]
```diff
@@ -105,7 +106,16 @@ int openmc_simulation_init()
...
+  }
+
+  // If fixed source and using custom source library then need to load
+  if (settings::run_mode == RunMode::FIXED_SOURCE &&
+        !settings::path_source_library.empty()) {
+	load_custom_source_library();
   }
```
Comment (simondrichards) [inline src/simulation.cpp]: Use two spaces per indent. I couldn't find mention in the OpenMC Style Guide for C++ about the indent for continuation lines, but elsewhere in the code it looks like the continuation line is aligned with the first character after the opening parenthesis on the preceding line.

```suggestion
  if (settings::run_mode == RunMode::FIXED_SOURCE &&
      !settings::path_source_library.empty()) {
    load_custom_source_library();
  }
```

**Comment by davidjohnlong** [comment]
The suggested changes have been made. Assigned review back to @simondrichards to accept.

File: src/state_point.cpp [hunk 4]
```diff
@@ -380,38 +380,42 @@ void load_state_point()
...
 
     // Take maximum of statepoint n_inactive and input n_inactive
     settings::n_inactive = std::max(settings::n_inactive, temp);
+
+    // Check to make sure source bank is present
+    if (settings::path_sourcepoint == settings::path_statepoint &&
+	    !source_present) {
+	  fatal_error("Source bank must be contained in statepoint restart file");
+    }
   }
```
Comment (simondrichards) [inline src/state_point.cpp]:
```suggestion
    if (settings::path_sourcepoint == settings::path_statepoint &&
        !source_present) {
      fatal_error("Source bank must be contained in statepoint restart file");
    }
  }

```

File: include/openmc/particle.h [hunk 5]
```diff
@@ -181,6 +181,15 @@ class Particle {
     int delayed_group; //!< particle delayed group
   };
 
+  //! Saved ("banked") state of primary source particle, for particle restart
+  struct RestartBank {
+    Position r;
+    Direction u;
+    double E;
+	double wgt;
+  };
```
Comment (simondrichards) [inline include/openmc/particle.h]: TAB character messing up indentation. Replace with spaces.

```suggestion
  struct RestartBank {
    Position r;
    Direction u;
    double E;
    double wgt;
  };
```

Comment (simondrichards) [inline include/openmc/particle.h]: TAB character messing up indentation. Replace with spaces.

```suggestion
  struct RestartBank {
    Position r;
    Direction u;
    double E;
    double wgt;
  };
```

File: src/simulation.cpp [hunk 6]
```diff
@@ -105,7 +106,16 @@ int openmc_simulation_init()
     load_state_point();
     write_message("Resuming simulation...", 6);
   } else {
-    initialize_source();
+    //Only initialize primary source bank for eigenvalue simulations
+    if (settings::run_mode == RunMode::EIGENVALUE) {
+	  initialize_source();
+    }
```
Comment (simondrichards) [inline src/simulation.cpp]:
```suggestion
    if (settings::run_mode == RunMode::EIGENVALUE) {
      initialize_source();
    }
```

File: src/simulation.cpp [hunk 7]
```diff
@@ -154,6 +164,12 @@ int openmc_simulation_finalize()
     t->active_ = false;
   }
 
+  // If fixed source and using custom source library then need to close
+  if (settings::run_mode == RunMode::FIXED_SOURCE &&
+      !settings::path_source_library.empty()) {
+  	close_custom_source_library();
+  }
```
Comment (simondrichards) [inline src/simulation.cpp]:
```suggestion
  if (settings::run_mode == RunMode::FIXED_SOURCE &&
      !settings::path_source_library.empty()) {
    close_custom_source_library();
  }
```

File: src/simulation.cpp [hunk 8]
```diff
@@ -429,18 +445,29 @@ void finalize_generation()
...
+	if (!settings::path_source_library.empty()) {
+	  site = sample_custom_source_library(&seed);
+	} else {
+	  site = sample_external_source(&seed);
+	}
+	p->from_source(&site);
+  } else if (settings::run_mode == RunMode::EIGENVALUE) {
+	// set defaults for eigenvalue simulations from primary bank
+    p->from_source(&simulation::source_bank[index_source - 1]);
+  }
```
Comment (simondrichards) [inline src/simulation.cpp]: More TABs messing up indentation: 

```suggestion
  if(settings::run_mode == RunMode::FIXED_SOURCE) {
    // initialize random number seed
    int64_t id = (simulation::total_gen + overall_generation() - 1)*settings::n_particles +
      simulation::work_index[mpi::rank] + index_source;
    uint64_t seed = init_seed(id, STREAM_SOURCE);
    // sample from external source distribution or custom library then set
    Particle::Bank site;
    if (!settings::path_source_library.empty()) {
      site = sample_custom_source_library(&seed);
    } else {
      site = sample_external_source(&seed);
    }
    p->from_source(&site);
  } else if (settings::run_mode == RunMode::EIGENVALUE) {
    // set defaults for eigenvalue simulations from primary bank
    p->from_source(&simulation::source_bank[index_source - 1]);
  }
```

**Review by simondrichards — CHANGES_REQUESTED** [review]
Sorry for the piecemeal review - I started this last night by adding individual comments, when I should have done it all as part of a proper review. A few more examples of wrong indentation, some of which are caused by TAB characters. It's probably better to avoid using TABs altogether and just use spaces.

Apart from the minor formatting issues it all looks good to me. It still passes the regression tests and no longer writes the source bank to the statepoint file, which is all good.

I noticed that the runtime section of the statepoint file still contains an entry for "sampling source sites" which now has the value zero. In some ways this makes sense as we no longer spend time filling the source bank. But the time for sampling source sites isn't zero in reality - it's just that it is now folded into the transport time, presumably. I'm not sure how we want to handle this - do we just leave this entry there with a value of zero, or remove this field entirely? It's probably not going to be practical or sensible to measure the times for sampling individual particles at transport time and summing them, is it? Maybe the best thing is to raise this question on the pull-request?

Once the minor formatting changes are committed I think this is ready to merge to your develop branch and submit a pull-request to openmc-dev/develop.

Thanks.

**Comment by davidjohnlong** [comment]
Further corrections to the formatting have been added. 

Agreed that the sampling time for the fixed source case is now "folded" into the transport time and it probably does not make sense to take individual timings of these and sum them over all samples, however this is a choice for the OpenMC community and as you state should be raised in the pull request.

File: src/simulation.cpp (hunk 8 above)
Comment (davidjohnlong) [reply src/simulation.cpp]: Done

File: src/simulation.cpp (hunk 7 above)
Comment (davidjohnlong) [reply src/simulation.cpp]: Done

File: src/simulation.cpp (hunk 6 above)
Comment (davidjohnlong) [reply src/simulation.cpp]: Done

**Comment by simondrichards** [comment]
Thanks @davidjohnlong - looks good.

**Pull request merged**